pkg/probe/http/request.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package http
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/component-base/version"
+	"k8s.io/kubernetes/pkg/probe"
+)
+
+// NewProbeRequest returns an http.Request suitable for use as a request for a
+// probe.
+func NewProbeRequest(url *url.URL, headers http.Header) (*http.Request, error) {
+	return newProbeRequest(url, headers, "probe")
+}
+
+// NewRequestForHTTPGetAction returns an http.Request derived from httpGet.
+// When httpGet.Host is empty, podIP will be used instead.
+func NewRequestForHTTPGetAction(httpGet *v1.HTTPGetAction, container *v1.Container, podIP string, userAgentFrament string) (*http.Request, error) {
+	scheme := strings.ToLower(string(httpGet.Scheme))
+	if scheme == "" {
+		scheme = "http"
+	}
+
+	host := httpGet.Host
+	if host == "" {
+		host = podIP
+	}
+
+	port, err := probe.ResolveContainerPort(httpGet.Port, container)
+	if err != nil {
+		return nil, err
+	}
+	path := httpGet.Path
+	url := formatURL(scheme, host, port, path)
+	headers := v1HeaderToHTTPHeader(httpGet.HTTPHeaders)
+
+	return newProbeRequest(url, headers, userAgentFrament)
+}
+
+func newProbeRequest(url *url.URL, headers http.Header, userAgentFragment string) (*http.Request, error) {
+	req, err := http.NewRequest("GET", url.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if headers == nil {
+		headers = http.Header{}
+	}
+	if _, ok := headers["User-Agent"]; !ok {
```

Comment:
I should have left comment https://github.com/kubernetes/kubernetes/pull/86139#discussion_r869769376 here instead - it still applies:

> I think there's a bug here, since headers are case-insensitive. I think buildHeader should use the header's Add method, which canonicalizes keys, and then use header.Get("User-Agent") here. Note that this is a slight behavior change though, which could affect probers...

Reply:
If you're worried about affecting probe behavior, you could put this behind the feature gate too.

Reply:
Yeah, I'd be worried about affecting probe behavior. Honestly, I'd rather save that for a future set of changes because this PR is not focused on changing probe behavior; it's about unifying lifecycle handler and probe behavior, and the probes currently don't use the `Get` method.

WDYT?

Reply:
Yeah, that makes sense. In that case can you leave a comment or TODO here?

Reply:
I added a clarifying comment; I think this is similar to what happens with the `Accept` header, where an explicitly set header with an empty value is treated differently than a header with an empty value.

Reply:
I don't see the comment - did you forget to push it?

Reply:
It's on line 72, right below this comment.

Reply:
Ah, I was picturing something more like this:
```suggestion
	// Note: This is applying case-sensitive matching to the case-insensitive header
	// key. This logic is incorrect, but correcting it would be a breaking-change, so
	// it is left as-is for now.
	if _, ok := headers["User-Agent"]; !ok {
```

Reply:
Oh, interesting. I thought this was more about presence than about casing. `Accept` is treated similarly.